File: membrain-py/tests/test_ollama.py
"""Tests for Ollama integration with Membrain.

These tests require a running Ollama instance at localhost:11434 with:
- nomic-embed-text (embedding model)
- llama3.1:8b or similar (chat model for extraction/conflict resolution)

Skip automatically if Ollama is not reachable.
"""

import time
import urllib.request
import urllib.error

import pytest

from membrain import MembrainClient


def _ollama_reachable() -> bool:
    """Check if Ollama is running at localhost:11434."""
    try:
        urllib.request.urlopen("http://localhost:11434/api/tags", timeout=2)
        return True
    except (urllib.error.URLError, OSError):
        return False


def _has_model(model_name: str) -> bool:
    """Check if a specific model is available in Ollama."""
    try:
        import json

        response = urllib.request.urlopen("http://localhost:11434/api/tags", timeout=2)
        data = json.loads(response.read())
        model_names = [m["name"].split(":")[0] for m in data.get("models", [])]
        return model_name in model_names
    except (urllib.error.URLError, OSError, ValueError):
        return False


requires_ollama = pytest.mark.skipif(
    not _ollama_reachable(),
    reason="Ollama not running at localhost:11434",
)

requires_embedding_model = pytest.mark.skipif(
    not _has_model("nomic-embed-text"),
    reason="nomic-embed-text model not available in Ollama",
)


def _chat_model() -> str:
    """Find the best available chat model in Ollama."""
    preferred = ["llama3.1", "llama3.2", "qwen2.5", "qwen3", "qwen3.5"]
    for model in preferred:
        if _has_model(model):
            # Return the full name with tag from the API
            try:
                import json

                response = urllib.request.urlopen(
                    "http://localhost:11434/api/tags", timeout=2,
                )
                data = json.loads(response.read())
                for m in data.get("models", []):
                    if m["name"].startswith(model):
                        return m["name"]
            except (urllib.error.URLError, OSError, ValueError):
                pass
    return ""


CHAT_MODEL = _chat_model()

requires_chat_model = pytest.mark.skipif(
    not CHAT_MODEL,
    reason="No suitable chat model available in Ollama",
)


def _make_embedding_client() -> MembrainClient:
    """Create a client with Ollama embeddings only."""
    return MembrainClient(config={
        "storage": {"backend": "memory"},
        "embedding": {
            "provider": "ollama",
            "model": "nomic-embed-text",
        },
    })


def _make_full_client() -> MembrainClient:
    """Create a client with Ollama embeddings + conflict resolution."""
    return MembrainClient(config={
        "storage": {"backend": "memory"},
        "embedding": {
            "provider": "ollama",
            "model": "nomic-embed-text",
        },
        "write": {
            "conflict_resolution": {
                "enabled": True,
                "provider": "ollama",
                "model": CHAT_MODEL,
            },
        },
    })


@requires_ollama
@requires_embedding_model
async def test_store_and_search_with_ollama_embeddings():
    """Store facts using Ollama embeddings and search for them."""
    client = _make_embedding_client()

    result = await client.store_fact(
        "Python was created by Guido van Rossum", confidence=0.9,
    )
    assert result.success, f"Store failed: {result.rejection_reason}"

    time.sleep(0.5)

    results = await client.search("Who created Python?", limit=5)
    assert len(results.memories) > 0
    contents = [m.content.lower() for m in results.memories]
    assert any("guido" in c for c in contents), f"Expected Guido in results: {contents}"

    client.close()


@requires_ollama
@requires_embedding_model
async def test_ollama_embedding_dimension():
    """Verify nomic-embed-text produces 768-dimensional embeddings."""
    client = _make_embedding_client()

    result = await client.store_fact("Test embedding dimension", confidence=0.9)
    assert result.success, f"Store failed: {result.rejection_reason}"

    client.close()


@requires_ollama
@requires_embedding_model
@requires_chat_model
async def test_conflict_resolution_with_ollama():
    """Full conflict resolution using Ollama for both embeddings and chat."""
    client = _make_full_client()

    result1 = await client.store_fact("Bob's favorite color is blue", confidence=0.9)
    assert result1.success, f"First store failed: {result1.rejection_reason}"

    time.sleep(1)

    result2 = await client.store_fact("Bob's favorite color is green", confidence=0.9)
    assert result2.success or result2.merged_with, (
        f"Second store failed: {result2.rejection_reason}"
    )

    time.sleep(1)

    results = await client.search("What is Bob's favorite color?", limit=5)
    contents = [m.content.lower() for m in results.memories]
    has_green = any("green" in c for c in contents)
    assert has_green, f"Expected 'green' in results, got: {contents}"

    client.close()


@requires_ollama
@requires_embedding_model
async def test_multiple_facts_with_ollama():
    """Store multiple facts and verify search relevance."""
    client = _make_embedding_client()

    facts = [
        "The Earth is the third planet from the Sun",
        "Water freezes at 0 degrees Celsius",
        "Light travels at approximately 300000 km per second",
    ]

    for fact in facts:
        result = await client.store_fact(fact, confidence=0.9)
        assert result.success, f"Store failed for '{fact}': {result.rejection_reason}"

    time.sleep(0.5)

    results = await client.search("What temperature does water freeze?", limit=5)
    assert len(results.memories) > 0
    top_content = results.memories[0].content.lower()
    assert "water" in top_content or "freeze" in top_content or "0" in top_content, (
        f"Expected water/freeze fact as top result, got: {top_content}"
    )

    client.close()

File: membrain-node/tests/test_ollama.mjs
/**
 * Tests for Ollama integration with Membrain.
 *
 * Requires a running Ollama instance at localhost:11434 with:
 * - nomic-embed-text (embedding model)
 * - llama3.1:8b or similar (chat model for conflict resolution)
 *
 * Skips automatically if Ollama is not reachable.
 */

import { describe, it } from "node:test";
import assert from "node:assert/strict";

import { MembrainClient } from "../dist/index.js";

async function ollamaReachable() {
  try {
    const response = await fetch("http://localhost:11434/api/tags", {
      signal: AbortSignal.timeout(2000),
    });
    return response.ok;
  } catch {
    return false;
  }
}

async function getModels() {
  try {
    const response = await fetch("http://localhost:11434/api/tags", {
      signal: AbortSignal.timeout(2000),
    });
    const data = await response.json();
    return (data.models || []).map((m) => m.name);
  } catch {
    return [];
  }
}

async function hasModel(name) {
  const models = await getModels();
  return models.some((m) => m.startsWith(name));
}

async function findChatModel() {
  const preferred = ["llama3.1", "llama3.2", "qwen2.5", "qwen3", "qwen3.5"];
  const models = await getModels();
  for (const prefix of preferred) {
    const found = models.find((m) => m.startsWith(prefix));
    if (found) return found;
  }
  return "";
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

const isReachable = await ollamaReachable();
const hasEmbedding = isReachable && (await hasModel("nomic-embed-text"));
const chatModel = isReachable ? await findChatModel() : "";

function makeEmbeddingClient() {
  return new MembrainClient({
    storage: { backend: "memory" },
    embedding: {
      provider: "ollama",
      model: "nomic-embed-text",
    },
  });
}

function makeFullClient() {
  return new MembrainClient({
    storage: { backend: "memory" },
    embedding: {
      provider: "ollama",
      model: "nomic-embed-text",
    },
    write: {
      conflict_resolution: {
        enabled: true,
        provider: "ollama",
        model: chatModel,
      },
    },
  });
}

describe("Ollama integration", () => {
  it("store and search with Ollama embeddings", { skip: !hasEmbedding }, async () => {
    const client = makeEmbeddingClient();
    try {
      const result = client.storeFact(
        "Python was created by Guido van Rossum",
        0.9
      );
      assert.ok(result.success, `Store failed: ${result.rejection_reason}`);

      await sleep(500);

      const results = client.search("Who created Python?", 5);
      assert.ok(results.memories.length > 0, "Expected search results");

      const contents = results.memories.map((m) => m.content.toLowerCase());
      const hasGuido = contents.some((c) => c.includes("guido"));
      assert.ok(hasGuido, `Expected Guido in results: ${contents}`);
    } finally {
      client.close();
    }
  });

  it("conflict resolution with Ollama", {
    skip: !hasEmbedding || !chatModel,
    timeout: 60000,
  }, async () => {
    const client = makeFullClient();
    try {
      const result1 = client.storeFact(
        "Bob's favorite color is blue",
        0.9
      );
      assert.ok(result1.success, `First store failed: ${result1.rejection_reason}`);

      await sleep(1000);

      const result2 = client.storeFact(
        "Bob's favorite color is green",
        0.9
      );
      assert.ok(
        result2.success || result2.merged_with,
        `Second store failed: ${result2.rejection_reason}`
      );

      await sleep(1000);

      const results = client.search("What is Bob's favorite color?", 5);
      const contents = results.memories.map((m) => m.content.toLowerCase());
      const hasGreen = contents.some((c) => c.includes("green"));
      assert.ok(hasGreen, `Expected 'green' in results, got: ${contents}`);
    } finally {
      client.close();
    }
  });

  it("multiple facts with Ollama embeddings", { skip: !hasEmbedding }, async () => {
    const client = makeEmbeddingClient();
    try {
      const facts = [
        "The Earth is the third planet from the Sun",
        "Water freezes at 0 degrees Celsius",
        "Light travels at approximately 300000 km per second",
      ];

      for (const fact of facts) {
        const result = client.storeFact(fact, 0.9);
        assert.ok(result.success, `Store failed for '${fact}': ${result.rejection_reason}`);
      }

      await sleep(500);

      const results = client.search("What temperature does water freeze?", 5);
      assert.ok(results.memories.length > 0, "Expected search results");

      const topContent = results.memories[0].content.toLowerCase();
      assert.ok(
        topContent.includes("water") || topContent.includes("freeze") || topContent.includes("0"),
        `Expected water/freeze as top result, got: ${topContent}`
      );
    } finally {
      client.close();
    }
  });
});
